File: modules/etude-widgets/android/src/main/java/expo/modules/etudewidgets/EtudeWidgetsModule.kt
package expo.modules.etudewidgets

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class WidgetData(
  @Field val today: Int = 0,
  @Field val goal: Int = 45,
  @Field val streak: Int = 0,
  @Field val week: List<Int> = emptyList(),
  @Field val nextFocus: String? = null,
  // #80: [accent, mid, soft] hex per scheme; empty = the brand terracotta from colors.xml
  @Field val accentLight: List<String> = emptyList(),
  @Field val accentDark: List<String> = emptyList(),
) : Record

/** Persists the widget snapshot and repaints placed widgets. */
object WidgetStore {
  private const val PREFS = "etude.widgets"

  fun save(ctx: Context, data: WidgetData) {
    ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
      .putInt("today", data.today)
      .putInt("goal", data.goal)
      .putInt("streak", data.streak)
      .putString("week", data.week.joinToString(","))
      .putString("nextFocus", data.nextFocus)
      .putString("accentLight", data.accentLight.joinToString(","))
      .putString("accentDark", data.accentDark.joinToString(","))
      .apply()
  }

  fun load(ctx: Context): WidgetData {
    val p = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    val week = p.getString("week", "")!!.split(',').mapNotNull { it.toIntOrNull() }
    return WidgetData(
      today = p.getInt("today", 0),
      goal = p.getInt("goal", 45),
      streak = p.getInt("streak", 0),
      week = week,
      nextFocus = p.getString("nextFocus", null),
      accentLight = p.getString("accentLight", "")!!.split(',').filter { it.isNotBlank() },
      accentDark = p.getString("accentDark", "")!!.split(',').filter { it.isNotBlank() },
    )
  }
}

class EtudeWidgetsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("EtudeWidgets")

    Function("setWidgetData") { data: WidgetData ->
      val ctx = appContext.reactContext ?: return@Function
      WidgetStore.save(ctx, data)
      updateAllWidgets(ctx)
    }
  }
}

File: modules/etude-widgets/ios/EtudeWidgetsModule.swift
import ExpoModulesCore
#if canImport(WidgetKit)
import WidgetKit
#endif

// Must match the App Group in targets/widgets/expo-target.config.js and app.json
let appGroup = "group.com.benstreich.etude"

struct WidgetData: Record {
  @Field var today: Int = 0
  @Field var goal: Int = 45
  @Field var streak: Int = 0
  @Field var week: [Int] = []
  @Field var nextFocus: String?
}

/**
 * Persists the widget snapshot into the shared App Group and asks WidgetKit
 * to repaint. The widget extension (targets/widgets) reads the same keys.
 */
public class EtudeWidgetsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("EtudeWidgets")

    Function("setWidgetData") { (data: WidgetData) in
      guard let defaults = UserDefaults(suiteName: appGroup) else { return }
      defaults.set(data.today, forKey: "today")
      defaults.set(data.goal, forKey: "goal")
      defaults.set(data.streak, forKey: "streak")
      defaults.set(data.week, forKey: "week")
      defaults.set(data.nextFocus, forKey: "nextFocus")
      #if canImport(WidgetKit)
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
      #endif
    }
  }
}
